crmsh/upgradeutil.py

```diff
@@ -0,0 +1,103 @@
+import logging
+import os.path
+import parallax
+import typing
+
+from . import parallax as crmparallax
+from . import remoteutil
+from . import utils as crmutils
+
+
+# pump this seq when upgrade check need to be run
+CURRENT_UPGRADE_SEQ = 1
+DATA_DIR = os.path.expanduser('~hacluster/crmsh')
+SEQ_FILE_PATH = DATA_DIR + '/upgrade_seq'
+FORCE_UPGRADE_FILE_PATH = DATA_DIR + '/upgrade_forced'
+
+
+logger = logging.getLogger(__name__)
+
+
+def _parallax_call(nodes, cmd):
+    parallax_options = parallax.Options()
+    parallax_options.ssh_options = ['StrictHostKeyChecking=no', 'ConnectTimeout=10']
+    return parallax.run(nodes, cmd, parallax_options)
+    # parallax.call raises exception if rc != 0, use an alternative implementation until it is fixed
+    #ret = dict()
+    #for node in nodes:
+    #    try:
+    #        result = remoteutil.run(node, cmd, use_shell=True)
+    #        result = (result.rc, result.stdout, result.stderr)
+    #    except remoteutil.Error as e:
+    #        logger.debug("Error calling %s on %s", cmd, node, exc_info=e)
+    #        result = e
+    #    ret[node] = result
+    #return ret
+
+
+def _upgrade_if_needed(nodes, fn: typing.Callable[[int], None]):
+    seq = 0
+    # FIXME: too slow, check localhost only
+    results = list(_parallax_call(nodes, 'test -f {}'.format(FORCE_UPGRADE_FILE_PATH)).values())
+    if any(isinstance(item, Exception) for item in results):
+        logger.info("parallax failed: nodes: %s, results: %s", nodes, results)
+        return
+    forced = any(rc == 0 for rc, _, _ in _parallax_call(nodes, 'test -f {}'.format(FORCE_UPGRADE_FILE_PATH)).values())
+    needed = forced
+    if not needed:
+        results = list(_parallax_call(nodes, 'cat {}'.format(SEQ_FILE_PATH)).values())
+        try:
+            seq = min(int(stdout.strip()) if rc == 0 else 0 for rc, stdout, stderr in results)
+        except ValueError:
+            seq = 0
+        needed = needed or seq < CURRENT_UPGRADE_SEQ
+    if needed:
+        fn(seq)
+        crmparallax.parallax_call(
+            nodes,
+            "mkdir -p '{data_dir}' && echo '{seq}' | cat > '{seq_file_path}'".format(
+                data_dir=DATA_DIR, seq=seq, seq_file_path=SEQ_FILE_PATH,
+            ),
+        )
+        crmparallax.parallax_call(nodes, "echo '{}' | cat > '{}'".format(CURRENT_UPGRADE_SEQ, SEQ_FILE_PATH))
+        crmparallax.parallax_call(nodes, 'rm -f {}'.format(FORCE_UPGRADE_FILE_PATH))
+
+
+def _upgrade(nodes, seq):
+    logger.info("Upgrading cluster...")
+    if seq == 0:
+        seq_0_setup_hacluster_passwordless(nodes)
+
+
+def upgrade_if_needed():
+    nodes = crmutils.list_cluster_nodes()
+    if nodes:
+        _upgrade_if_needed(nodes, lambda seq: _upgrade(nodes, seq))
```

Comment:
Given a 5 nodes cluster, what's the upgrade sequence would be, node by node from scratch? It reasonable to deal with those upgraded nodes, but probably too aggressive to deal with those old nodes?

Reply:
The sequence is run `crm cluster init ssh -y` on one of the nodes and `crm cluster join ssh  -c {} -y` on others. Do you mean we need to deal with a partially upgraded cluster, with nodes running different versions of ha stack?

Reply:
> The sequence is run `crm cluster init ssh -y` on one of the nodes and `crm cluster join ssh -c {} -y` on others. Do you mean we need to deal with a partially upgraded cluster, with nodes running different versions of ha stack?

What you described here is to construct a brand new cluster, and even you may assume all nodes are upgraded already.  However, the use case scenario of this PR is more complex and it is to upgrade an existing cluster, and some nodes are upgraded and some might not yet.

Reply:
OK. I will delay the ssh init/join sequence until all nodes have been upgraded to new version.

Reply:
In regarding to the version upgrade, I'm not sure if crmsh could leverage the bigger picture from pacemaker, for example,
https://clusterlabs.org/pacemaker/doc/2.1/Pacemaker_Administration/html/upgrading.html#rolling-node-by-node